File: src/data/courses/lld/real-world-problems/uber-favorites/design-logging-framework/code/java/LoggingFramework.java
// DESIGN: Logging Framework — LLD (Java) | CoR + Strategy + Singleton
import java.util.*;
import java.time.*;
import java.time.format.*;
enum LogLevel { DEBUG, INFO, WARN, ERROR, FATAL }
class LogMessage { LogLevel level; String message, timestamp;
    LogMessage(LogLevel l,String m){level=l;message=m;timestamp=LocalDateTime.now().format(DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss"));}
}
interface LogFormatter { String format(LogMessage msg); }
class PlainFmt implements LogFormatter { public String format(LogMessage m){return "["+m.timestamp+"] ["+m.level+"] "+m.message;} }
class JSONFmt implements LogFormatter { public String format(LogMessage m){return "{ts:"+m.timestamp+",level:"+m.level+",msg:"+m.message+"}";} }
abstract class LogHandler {
    LogLevel minLevel; LogHandler next; LogFormatter formatter;
    LogHandler(LogLevel min,LogFormatter fmt){minLevel=min;formatter=fmt;}
    LogHandler setNext(LogHandler h){next=h;return h;}
    void handle(LogMessage msg){if(msg.level.ordinal()>=minLevel.ordinal())write(formatter.format(msg));if(next!=null)next.handle(msg);}
    abstract void write(String formatted);
}
class ConsoleHandler extends LogHandler {
    ConsoleHandler(LogLevel min,LogFormatter f){super(min,f);} void write(String m){System.out.println(m);}
}
class FileHandler extends LogHandler {
    String file; FileHandler(LogLevel min,LogFormatter f,String file){super(min,f);this.file=file;}
    void write(String m){System.out.println("[FILE:"+file+"] "+m);}
}
class Logger {
    private static Logger instance; LogHandler chain;
    private Logger(){} static Logger getInstance(){if(instance==null)instance=new Logger();return instance;}
    void setChain(LogHandler h){chain=h;}
    void log(LogLevel l,String m){if(chain!=null)chain.handle(new LogMessage(l,m));}
    void debug(String m){log(LogLevel.DEBUG,m);} void info(String m){log(LogLevel.INFO,m);}
    void warn(String m){log(LogLevel.WARN,m);} void error(String m){log(LogLevel.ERROR,m);}
}
public class LoggingFramework {
    public static void main(String[] args) {
        System.out.println("=== Logging — Java ===");
        ConsoleHandler con=new ConsoleHandler(LogLevel.DEBUG,new PlainFmt());
        FileHandler fil=new FileHandler(LogLevel.INFO,new JSONFmt(),"app.log");
        con.setNext(fil);
        Logger.getInstance().setChain(con);
        Logger.getInstance().debug("Starting...");
        Logger.getInstance().info("Server up");
        Logger.getInstance().error("Timeout!");
        System.out.println("=== Complete ===");
    }
}
// SUMMARY: CoR(console->file, ALWAYS passes), Strategy(formatters), Singleton(logger)

File: src/data/courses/lld/real-world-problems/uber-favorites/design-logging-framework/code/cpp/LoggingFramework.cpp
// ===========================================================================
// DESIGN: Logging Framework — Low Level Design
// Patterns: Chain of Responsibility, Strategy, Singleton
// ===========================================================================
// FR: Log levels (DEBUG→FATAL), multiple outputs, formatters, level filtering
// PLANTUML: Logger(Singleton)*--LogHandler(chain), LogHandler-->LogFormatter(Strategy)

#include <iostream>
#include <string>
#include <ctime>
#include <cstdio>
using namespace std;

enum LogLevel { LOG_DEBUG=0, LOG_INFO=1, LOG_WARN=2, LOG_ERROR=3, LOG_FATAL=4 };

string levelStr(LogLevel l) {
    switch(l) { case LOG_DEBUG:return"DEBUG"; case LOG_INFO:return"INFO"; case LOG_WARN:return"WARN";
        case LOG_ERROR:return"ERROR"; case LOG_FATAL:return"FATAL"; }
    return"";
}

struct LogMessage {
    LogLevel level;
    string message, timestamp;
    LogMessage(LogLevel l, const string& m) : level(l), message(m) {
        time_t now = time(0);
        char buf[20];
        strftime(buf, sizeof(buf), "%Y-%m-%d %H:%M:%S", localtime(&now));
        timestamp = string(buf);
    }
};

// Strategy: Formatter
class LogFormatter {
public:
    virtual ~LogFormatter() {}
    virtual string format(const LogMessage& msg) const = 0;
};

class PlainFormatter : public LogFormatter {
public:
    string format(const LogMessage& msg) const {
        return "[" + msg.timestamp + "] [" + levelStr(msg.level) + "] " + msg.message;
    }
};

class JSONFormatter : public LogFormatter {
public:
    string format(const LogMessage& msg) const {
        return "{\"ts\":\"" + msg.timestamp + "\",\"level\":\"" + levelStr(msg.level) + "\",\"msg\":\"" + msg.message + "\"}";
    }
};

// Chain of Responsibility: Handler
class LogHandler {
protected:
    LogLevel minLevel;
    LogHandler* next;
    LogFormatter* formatter;
public:
    LogHandler(LogLevel min, LogFormatter* fmt) : minLevel(min), next(NULL), formatter(fmt) {}
    virtual ~LogHandler() {}
    void setNext(LogHandler* h) { next = h; }

    void handle(const LogMessage& msg) {
        if ((int)msg.level >= (int)minLevel) write(formatter->format(msg));
        if (next) next->handle(msg);  // ALWAYS pass to next (unlike approval chain)
    }

    virtual void write(const string& formatted) = 0;
};

class ConsoleHandler : public LogHandler {
public:
    ConsoleHandler(LogLevel min, LogFormatter* fmt) : LogHandler(min, fmt) {}
    void write(const string& msg) { cout << msg << endl; }
};

class FileHandler : public LogHandler {
    string filename;
public:
    FileHandler(LogLevel min, LogFormatter* fmt, const string& file)
        : LogHandler(min, fmt), filename(file) {}
    void write(const string& msg) { cout << "[FILE:" << filename << "] " << msg << endl; }
};

// Singleton Logger
class Logger {
    LogHandler* chain;
    Logger() : chain(NULL) {}
    Logger(const Logger&);
    static Logger* instance;
public:
    static Logger* getInstance() {
        if (!instance) instance = new Logger();
        return instance;
    }
    void setChain(LogHandler* h) { chain = h; }
    void log(LogLevel level, const string& msg) { if(chain) chain->handle(LogMessage(level, msg)); }
    void debug(const string& m) { log(LOG_DEBUG, m); }
    void info(const string& m)  { log(LOG_INFO, m); }
    void warn(const string& m)  { log(LOG_WARN, m); }
    void error(const string& m) { log(LOG_ERROR, m); }
    void fatal(const string& m) { log(LOG_FATAL, m); }
};
Logger* Logger::instance = NULL;

int main() {
    cout << "============================================\n  Logging Framework — LLD Demo\n============================================" << endl;

    PlainFormatter plain;
    JSONFormatter json;
    ConsoleHandler console(LOG_DEBUG, &plain);  // Console: all levels
    FileHandler file(LOG_INFO, &json, "app.log"); // File: INFO+
    console.setNext(&file);  // Chain: Console -> File

    Logger::getInstance()->setChain(&console);
    Logger::getInstance()->debug("App starting...");     // Console only
    Logger::getInstance()->info("Server listening");      // Console + File
    Logger::getInstance()->warn("High memory usage");     // Console + File
    Logger::getInstance()->error("Connection timeout");   // Console + File
    Logger::getInstance()->fatal("System crash!");         // Console + File

    cout << "\n=== Complete ===" << endl;
    return 0;
}
// SUMMARY: CoR(handler chain processes AND passes), Strategy(formatters), Singleton(logger)
// KEY: Logging chain ALWAYS passes (unlike approval chain that stops)
// PRACTICES: SRP per handler, OCP(new handlers/formatters), DIP(Logger->LogHandler abstraction)
